Moving `pallet-asset-tx-payment` from cumulus to substrate

georgesdib opened the pull request:
Fixes paritytech/cumulus#688.

Moving `pallet-asset-tx-payment` into substrate given there is nothing cumulus specific in it.

@apopiak @NukeManDan 
related cumulus PR building on this one: paritytech/cumulus#712

apopiak commented:
Thank you for your contribution! I took the liberty of moving the pallet into `transaction-payment` because it is an extension of that pallet.

apopiak commented:
@shawntabrizi Do we want to add this pallet to the node runtime? Will break transaction format, but would probably be good to make sure that it keeps working with any changes.

bkchr commented:
The Substrate node is just for testing, for nothing else. So, it should be fine to add this.

apopiak commented:
@georgesdib 
Could you please add the new pallet to the node runtime

georgesdib commented:
> @georgesdib
> Could you please add the new pallet to the node runtime

Done, however I had to do quite a few changes, and some (like changing `Balance` from `u64` to `u128`) could be controversial.

apopiak commented on bin/node/cli/src/service.rs:
```diff
@@ -719,6 +721,7 @@ mod tests {
 				let check_nonce = frame_system::CheckNonce::from(index);
 				let check_weight = frame_system::CheckWeight::new();
 				let payment = pallet_transaction_payment::ChargeTransactionPayment::from(0);
```
This needs to be removed. `ChargeAssetTxPayment` replaces `ChargeTransactionPayment`

apopiak replied:
```suggestion
```

apopiak commented on bin/node/runtime/src/lib.rs:
```diff
@@ -969,6 +977,7 @@ where
 			frame_system::CheckNonce::<Runtime>::from(nonce),
 			frame_system::CheckWeight::<Runtime>::new(),
 			pallet_transaction_payment::ChargeTransactionPayment::<Runtime>::from(tip),
```
```suggestion
```

apopiak commented on bin/node/runtime/src/lib.rs:
```diff
@@ -1315,6 +1325,7 @@ pub type SignedExtra = (
 	frame_system::CheckNonce<Runtime>,
 	frame_system::CheckWeight<Runtime>,
 	pallet_transaction_payment::ChargeTransactionPayment<Runtime>,
```
```suggestion
```

apopiak commented on bin/node/runtime/src/lib.rs:
```diff
@@ -1256,6 +1265,7 @@ construct_runtime!(
 		Indices: pallet_indices,
 		Balances: pallet_balances,
 		TransactionPayment: pallet_transaction_payment,
```
Just as a note: this stays, we keep the pallet, just the SignedExtension is replaced.

apopiak commented on bin/node/test-runner-example/src/lib.rs:
```diff
@@ -78,6 +78,7 @@ impl ChainInfo for NodeTemplateChainInfo {
 			),
 			frame_system::CheckWeight::<Self::Runtime>::new(),
 			pallet_transaction_payment::ChargeTransactionPayment::<Self::Runtime>::from(0),
```
```suggestion
```

apopiak reviewed (commented):
The SignedExtension is replaced, not added.

apopiak commented:
@jacogr I imagine this PR would have an impact on how Polkadot-js handles the default Substrate node.

jacogr commented:
No impact, the metadata defined which extensions should be used. (As long as it knows the type and in this case it does... assuming no changes to the format of the extension...)

apopiak commented:
Ah, ok. The extension should be the same as the one in Cumulus/Statemint (we're just moving repos after all). :+1:

apopiak commented:
@georgesdib Some tests are failing, can you see them?

georgesdib commented:
> @georgesdib Some tests are failing, can you see them?

Yes, I will work on this. I don't fully get the reason but yes I see the error.

georgesdib commented:
> @georgesdib Some tests are failing, can you see them?

I am not sure of myself, but debugging the failing test, it seems to be occurring when `pallet_timestamp::set` is being called to set the time. This is an unsigned extrinsic, so the execution goes to `pre_dispatch_unsigned` which creates a default `Pre`.

The default `Pre` has `InitialPayment::Nothing` set, but the `post_info` says that it pays fees. This combination in the function `post_dispatch` of the pallet `asset_tx_payment` fails the debug assert.

It seems to be that:

- Either the fact that `pallet_timestamp::set` seems to return that it pays fee is wrong (which I would put as very very unlikely)
- Or the `pre_dispatch_unsigned` needs to be overridden in `pallet_asset_tx_payment` to not return `InitialPayment::Nothing`, which seems like the plausible best fix.
- Or, amend `post_dispatch` to remove the debug assert, but I suspect that it's there for a valid reason.
- Or, I am simply wrong in my assessment :)

georgesdib commented:
> > @georgesdib Some tests are failing, can you see them?
> 
> 
> 
> I am not sure of myself, but debugging the failing test, it seems to be occurring when `pallet_timestamp::set` is being called to set the time. This is an unsigned extrinsic, so the execution goes to `pre_dispatch_unsigned` which creates a default `Pre`.
> 
> 
> 
> The default `Pre` has `InitialPayment::Nothing` set, but the `post_info` says that it pays fees. This combination in the function `post_dispatch` of the pallet `asset_tx_payment` fails the debug assert.
> 
> 
> 
> It seems to be that:
> 
> 
> 
> - Either the fact that `pallet_timestamp::set` seems to return that it pays fee is wrong (which I would put as very very unlikely)
> 
> - Or the `pre_dispatch_unsigned` needs to be overridden in `pallet_asset_tx_payment` to not return `InitialPayment::Nothing`, which seems like the plausible best fix.
> 
> - Or, amend `post_dispatch` to remove the debug assert, but I suspect that it's there for a valid reason.
> 
> - Or, I am simply wrong in my assessment :)

I debugged putting `transaction-payment` back and I would say this is consistent with removing the `debug_assert` on the fee. But I assume it's there for a reason so not sure what I am getting wrong.

@apopiak thoughts?

apopiak commented:
Well, I put the `debug_assert` there because AFAICT my code (and the original transaction payment pallet) assume that the amount of fees never increases from pre- to post-dispatch.

apopiak commented:
related to #10143

apopiak commented:
Aaah, `pre_dispatch_unsigned` returns fees of zero but is *not*  `Pays::No` I imagine.

kianenigma commented on bin/node/runtime/src/impls.rs:
```diff
@@ -27,6 +31,18 @@ impl OnUnbalanced<NegativeImbalance> for Author {
+impl HandleCredit<AccountId, Assets> for CreditToBlockAuthor {
+	fn handle_credit(credit: CreditOf<AccountId, Assets>) {
+		let author = pallet_authorship::Pallet::<Runtime>::author();
+		// TODO: what to do in case paying the author fails (e.g. because `fee < min_balance`)
```
TODO should either be removed or linked to an issue.

georgesdib commented on frame/transaction-payment/asset-tx-payment/src/lib.rs:
```diff
@@ -0,0 +1,299 @@
+		let (tip, who, initial_payment) = pre;
+		match initial_payment {
+			InitialPayment::Native(already_withdrawn) => {
+				let actual_fee = pallet_transaction_payment::Pallet::<T>::compute_actual_fee(
```
on it

apopiak commented on frame/transaction-payment/asset-tx-payment/src/lib.rs:
```diff
@@ -255,21 +248,19 @@ where
-					info,
-					post_info,
-					actual_fee,
+				pallet_transaction_payment::ChargeTransactionPayment::<T>::do_post_dispatch(
```
Should not be necessary if you construct `ChargeTransactionPayment::Pre` (as `(tip, who, already_withdrawn)`), but this is fine as well I would say :shrug:

georgesdib replied:
Refactored as suggested, as much simpler indeed.

apopiak commented on frame/transaction-payment/src/lib.rs:
```diff
@@ -679,6 +679,22 @@ where
 		.saturated_into::<TransactionPriority>()
 	}
+
+	pub fn do_post_dispatch(
```
You will need to document this, though, if you introduce another `pub` function

georgesdib replied:
Not needed anymore.

georgesdib commented:
Checks seem to have timed out, is it possible to kick it off again without a dummy commit?

apopiak commented:
It shows all checks as successful to me

georgesdib commented:
> It shows all checks as successful to me

Same for me too now, odd, but good news :)

shawntabrizi commented on frame/transaction-payment/asset-tx-payment/src/tests.rs:
```diff
@@ -0,0 +1,748 @@
+impl HandleCredit<AccountId, Assets> for CreditToBlockAuthor {
+	fn handle_credit(credit: CreditOf<AccountId, Assets>) {
+		let author = pallet_authorship::Pallet::<Runtime>::author();
+		// TODO: what to do in case paying the author fails (e.g. because `fee < min_balance`)
+		// default: drop the result which will trigger the `OnDrop` of the imbalance.
```
todo

georgesdib replied:
@apopiak That is the same as what @kianenigma had commented and you removed the `TODO`. I assume the best here is simply to remove the `TODO` word and leave the rest?

apopiak replied:
Yeah, just remove it. The default behavior is fine.

shawntabrizi reviewed (approved):
looks reasonable to me

shawntabrizi commented:
We should probably have one more review on top of myself and alex since alex already played a large role in this pr

georgesdib commented:
I seemed to have uncovered a breaking change there.
`cargo update` updated `impl-serde` to `0.3.2` from `0.3.1`
The function `from_hex` of that version has been "fixed" to be able to parse strings that do not start with "0x", while it did not parse this before.
The failing test is expecting the old behaviour and is therefore failing.

The easy solution would be to just revert back my `cargo.lock` change, but I think this is worth pondering if it's the right behaviour or not?

georgesdib commented:
> How is this related to this PR?

It's not, I have reverted the `cargo.lock` change.

> How is this related to this PR?

It's not. I have reverted the `impl-serde` accidental upgrade.

georgesdib commented:
All checks are green, @kianenigma @tomusdrw look good to merge?

kianenigma commented on frame/transaction-payment/asset-tx-payment/src/payment.rs:
```diff
@@ -0,0 +1,168 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+///! Traits and default implementation for paying transaction fees in assets.
```
```suggestion
///! Traits and default implementation for paying transaction fees in assets.

```

georgesdib replied:
This actually broke cargo fmt so had to revert it

kianenigma reviewed (approved):
Looks good to me!

georgesdib commented:
@apopiak we have all approvals, shall we merge it?

apopiak commented:
bot merge

bkchr commented:
/tip small

bkchr commented:
@georgesdib if you add your address, I can ping the bot another time.

georgesdib commented:
> @georgesdib if you add your address, I can ping the bot another time.

Done, thank you very much!

bkchr commented:
/tip small
